SectionedSurface Aplitop 1

javierNadal opened the pull request:
## Unit test

This scenario is a simple road visualization (IfcSectionedSurface and IfcOpenCrossProfileDef)

## Covered topics

*tick one/many*

- [ ] project setup
- [ ] georeferencing
- [x] alignment semantics
- [ ] alignment geometry
- [ ] linear placement
- [ ] project breakdown structure / spatial structure
- [ ] terrain
- [ ] triangulated irregular network
- [x] open cross profile
- [ ] sectioned surface & solid
- [ ] pavement & course
- [ ] surface feature
- [ ] earthworks
- [ ] geotechnics
- [ ] longitudinal products (barriers, guardrail, ...)
- [ ] interchange
- [ ] structural
- [ ] drainage
- [ ] signage
- [ ] road furniture
- [ ] rail furniture
- [ ] ports & waterways furniture
- [ ] bridge furniture

javierNadal committed:
Last alignment changes pending (Meeting Feb 10th)

javierNadal committed:
Alignment-Aplitop-1 RC3

javierNadal committed:
readme corrections

javierNadal committed:
More readme corrections

javierNadal committed:
Merge pull request #1 from bSI-InfraRoom/main

Updating changes

javierNadal committed:
Updated with latest schema (12 March 2021)

javierNadal committed:
Wrong file

javierNadal committed:
Merge pull request #2 from bSI-InfraRoom/main

Updating

javierNadal committed:
Merge pull request #3 from bSI-InfraRoom/main

Updating

javierNadal committed:
SectionedSurface-Aplitop-1

javierNadal committed:
Updated png and ifc files

larswik commented on SectionedSurface-Aplitop-1/SectionedSurface-Aplitop-1.IFC:
```diff
@@ -0,0 +1,93 @@
+#45=IFCRELNESTS('1qVzEdro5CFf5K20Wsg4rv',#9,$,$,#31,(#33));
+#46=IFCCOMPOSITECURVE((#44),.F.);
+#47=IFCGEOMETRICREPRESENTATIONSUBCONTEXT('Axis','Model',*,*,*,*,#19,$,.GRAPH_VIEW.,$);
+#48=IFCSHAPEREPRESENTATION(#47,'Axis','Curve2D',(#46));
```
When there is an IfcGradientCurve available, I wonder if the shape representation for an IfcAlignment should refer to this  (65) instead if the IfcCompositeCurve?

javierNadal replied:
In a simple sample like this, just a straight line, yes. 

But in a more general sample, I still don't have an implementation to build a single IfcCompositeCurve from any 2D alingment and grade line info. So I kept alignment representation as 2D.

I will change it.

larswik replied:
It seems as if the IfcProductDefinitionShape on line #49 no longer has an owning IfcProduct. I assume that this should be the IfcAlignment?

larswik replied:
I see that IfcAlignmentHorizontal has a shape representation = IfcCompositeCurve.
IfcAlignmentVertical has a shape representation = IfcGradientCurve
I wonder whether IfcAlignment (no shape representation in this file) should have the gradient curve as its representation (Axis,Curve3D)? @SergejMuhic ?

larswik commented on SectionedSurface-Aplitop-1/SectionedSurface-Aplitop-1.IFC:
```diff
@@ -0,0 +1,93 @@
+
+#272=IFCPRODUCTDEFINITIONSHAPE($,$,(#273));
+#273=IFCSHAPEREPRESENTATION(#19,'Profiles','',(#274));
+#274=IFCSECTIONEDSURFACE(#65,(#275,#277,#279,#281,#283),(#276,#278,#280,#282,#284),.F.);
```
It looks to me as if the list of CrossSectionPositions should come as attribute 2 and CrossSections as attribute nr 3. I checked with the KIT checker where the data passes regardless of the ordering between these two?

javierNadal replied:
Yes, you are right. I will swap them.

larswik commented on SectionedSurface-Aplitop-1/SectionedSurface-Aplitop-1.IFC:
```diff
@@ -0,0 +1,93 @@
+#274=IFCSECTIONEDSURFACE(#65,(#275,#277,#279,#281,#283),(#276,#278,#280,#282,#284),.F.);
+
+#275=IFCOPENCROSSPROFILEDEF(.CURVE.,$,.T.,(0.031,3.5,3.5,0.056),(-0.667,0.,0.,-0.667),('119','30','0','30','139'));
+#276=IFCPOINTBYDISTANCEEXPRESSION(IFCLENGTHMEASURE(0.),IFCLENGTHMEASURE(3.531),0.,0.,#65);
```
From looking at the xml data, it would seem that there should be a negative vertical offset needed in this case. Also, i am not sure where the value 3.531 comes from. In the xml at station 0 we have -2.736 and at station 20 we have -3.778.

javierNadal replied:
Value 3.351 is a left displacement of the OpenCrossProfile. First two widths (0.031 & 3.5) are located at left side of alignment.
I will check vertical offsets.

larswik commented on SectionedSurface-Aplitop-1/SectionedSurface-Aplitop-1.IFC:
```diff
@@ -0,0 +1,93 @@
+#273=IFCSHAPEREPRESENTATION(#19,'Profiles','',(#274));
+#274=IFCSECTIONEDSURFACE(#65,(#275,#277,#279,#281,#283),(#276,#278,#280,#282,#284),.F.);
+
+#275=IFCOPENCROSSPROFILEDEF(.CURVE.,$,.T.,(0.031,3.5,3.5,0.056),(-0.667,0.,0.,-0.667),('119','30','0','30','139'));
```
The slope values seem off. Please consult the documentation (x to the left and y up, positive angles from x to y). Typically, the slope value when starting from the left would be relatively close to PI radians.

Also, unique tags would be recommended.

javierNadal replied:
Yes, I'm changing slopes

larswik commented on SectionedSurface-Aplitop-1/SectionedSurface-Aplitop-1.IFC:
```diff
@@ -0,0 +1,93 @@
+#277=IFCOPENCROSSPROFILEDEF(.CURVE.,$,.T.,(0.543,3.5,3.5,0.301),(-0.6667,0.,0.,-0.667),('119','30','0','30','119'));
+#278=IFCPOINTBYDISTANCEEXPRESSION(IFCLENGTHMEASURE(30.),IFCLENGTHMEASURE(4.043),0.,0.,#65);
+
+#279=IFCOPENCROSSPROFILEDEF(.CURVE.,$,.T.,(0.663,3.5,3.5,3.154),(-0.667,0.,0.,0.6667),('139','30','0','30','139'));
```
If it is by design to switch from tag '119' to tag '139', I would recommend adding '139' to the previous cross section or '119' to this. See documentation:
![image](https://user-images.githubusercontent.com/25506404/120165841-421c4380-c1fc-11eb-9b1a-413069bacd3d.png)

javierNadal replied:
Yes, I will include one more vertex with 139 code in previous/next profile, and 0 width vector.

javierNadal committed:
Some corrections

larswik commented on SectionedSurface-Aplitop-1/SectionedSurface-Aplitop-1.IFC:
```diff
@@ -0,0 +1,87 @@
+#46=IFCCOMPOSITECURVE((#44),.F.);
+#47=IFCGEOMETRICREPRESENTATIONSUBCONTEXT('Axis','Model',*,*,*,*,#19,$,.GRAPH_VIEW.,$);
+#48=IFCSHAPEREPRESENTATION(#47,'Axis','Curve2D',(#46));
+#49=IFCPRODUCTDEFINITIONSHAPE($,$,(#48));
```
This line should be referenced somewhere.

javierNadal replied:
In updated file, #49 is representation of IFCHORIZONTALALINGMENT

larswik commented on SectionedSurface-Aplitop-1/SectionedSurface-Aplitop-1.IFC:
```diff
@@ -0,0 +1,87 @@
+#65=IFCGRADIENTCURVE((#63),.F.,#46,$);
+#66=IFCSHAPEREPRESENTATION(#47,'Axis','Curve3D',(#65));
+#67=IFCPRODUCTDEFINITIONSHAPE($,$,(#66));
+#68=IFCFACILITYPART('2VJRykc916eO0F0TUHiUj9',$,$,$,'ROADSEGMENT',#24,#87,$,.COMPLEX.,$,.LONGITUDINAL.);
```
There's an error here. The 10th (penutimate) attribute should be the predefined type. Suggestion: IFCROADPARTTYPEENUM(.ROADSEGMENT.). The 5th attribute (ObjectType) can thereby be removed (set to $) since it doesn't add anything.

javierNadal replied:
Done

larswik commented on SectionedSurface-Aplitop-1/SectionedSurface-Aplitop-1.IFC:
```diff
@@ -0,0 +1,87 @@
+ISO-10303-21;
+HEADER;
+FILE_DESCRIPTION(('ViewDefinition []'),'2;1');
```
ViewDefinition can be set to [Ifc4X3NotAssigned] for now.

javierNadal replied:
Done

javierNadal committed:
Some corrections

Fixed IFCROADPARTTYPEENUM(.ROADSEGMENT.) , representation of IFCALIGNMENTHORIZONTAL, and ViewDefinition

larswik commented on SectionedSurface-Aplitop-1/SectionedSurface-Aplitop-1.IFC:
```diff
@@ -0,0 +1,87 @@
+#72=IFCPOINTBYDISTANCEEXPRESSION(IFCLENGTHMEASURE(0.),3.531,0.,0.,#65);
+#73=IFCOPENCROSSPROFILEDEF(.CURVE.,'CrossProfile-2',.T.,(0.543,3.5,3.5,0.301),(2.47492598733333,3.141592654,3.141592654,2.47714082675747),('119','30','0','30','119'));
+#75=IFCPOINTBYDISTANCEEXPRESSION(IFCLENGTHMEASURE(30.),4.043,0.,0.,#65);
+#76=IFCOPENCROSSPROFILEDEF(.CURVE.,'CrossProfile-3',.T.,(0.663,3.5,3.5,0.,3.154),(2.47492598733333,3.141592654,3.141592654,3.141592654,3.80836500656817),('119','30','0','30','119','139'));
```
My interpretation is according to this image:
![image](https://user-images.githubusercontent.com/25506404/122394428-42e00400-cf76-11eb-9c41-365f2e6bba97.png)
Is it possible that cross section at 60 should join 119 and 139 (at section 60) instead of 119 and 30, i.e. have the width 0 as last value?
Please also check that my interpretation is correct. Other than that, I think that the file LFTM!

javierNadal replied:
I have added a new IfcOpenCrossProfileDef at station 45. Now I think it looks better.
![image](https://user-images.githubusercontent.com/69001750/122417854-1504ba80-cf8a-11eb-8c10-462321007d49.png)

javierNadal replied:
I have added a new IfcOpenCrossProfileDef at station 45. Now I think it looks better.

![image](https://user-images.githubusercontent.com/69001750/122417957-2e0d6b80-cf8a-11eb-96d7-8076eb728c6a.png)

javierNadal committed:
Added new IfcOpenCrossProfileDef at station 45

jmirtsch commented on SectionedSurface-Aplitop-1/SectionedSurface-Aplitop-1.IFC:
```diff
@@ -0,0 +1,92 @@
+#68=IFCFACILITYPART('2VJRykc916eO0F0TUHiUj9',$,$,$,$,#24,#87,$,.COMPLEX.,IFCROADPARTTYPEENUM(.ROADSEGMENT.),.LONGITUDINAL.);
+#69=IFCRELAGGREGATES('0nVIlPiJ1D0Opi4EVcPTpM',#9,$,$,#28,(#68));
+#70=IFCOPENCROSSPROFILEDEF(.CURVE.,'CrossProfile-1',.T.,(0.0310000000000001,3.5,3.5,0.056),(2.49643136367755,3.141592654,3.141592654,2.48087836828573),('119','30','0','30','119'));
+#72=IFCPOINTBYDISTANCEEXPRESSION(IFCLENGTHMEASURE(0.),3.531,0.,0.,#65);
```
KIT checker doesn't seem to detect this, but shouldn't DistanceAlong be IfcNonNegativeLengthMeasure
 and not IfcLengthMeasure ??  This applies too all occurrences.

larswik replied:
It should be an IfcNonNegativeLengthMeasure for the IfcPointByDistanceExpression:
TYPE IfcCurveMeasureSelect = SELECT (
 IfcParameterValue,
 IfcNonNegativeLengthMeasure);
END_TYPE;

larswik replied:
One more thing. I discussed with the rest of the team this morning and even if the documentation lacks this information, the conclusion was that it is at least recommended to keep the tags unique within a profile. Would it be possible to e.g. add a "R" and "L" prefix to separate e.g. 30 on the left vs 30 on the right?

javierNadal replied:
Done

javierNadal committed:
L&R prefixes in codes. IFCNONNEGATIVELENGTHMEASURE instead of IFCLENGTHMEASURE

jmirtsch commented on SectionedSurface-Aplitop-1/SectionedSurface-Aplitop-1.IFC:
```diff
@@ -0,0 +1,92 @@
+#33=IFCALIGNMENTSEGMENT('0sVreI8Kn6tuw4UYDmeHtA',$,$,$,$,#24,$,#35);
+#34=IFCCARTESIANPOINT((286.47559897,239.79766567));
+#35=IFCALIGNMENTHORIZONTALSEGMENT($,$,#34,1.50346536986339,0.,0.,107.15812752,$,.LINE.);
+#36=IFCDIRECTION((0.999655739909883,0.0262374096515884));
```
Semantics of Alignment seem to be south to north, where as this direction positions the geometry to near east/west (ie x).  Which is correct?
> IFCDIRECTION((0.999655739909883,0.0262374096515884));

![210624 alignment direction gg](https://user-images.githubusercontent.com/791757/123183956-03fe0100-d4d6-11eb-8802-502e01a56433.png)

jmirtsch commented on SectionedSurface-Aplitop-1/SectionedSurface-Aplitop-1.IFC:
```diff
@@ -0,0 +1,92 @@
+#72=IFCPOINTBYDISTANCEEXPRESSION(IFCNONNEGATIVELENGTHMEASURE(0.),3.531,0.,0.,#65);
+#73=IFCOPENCROSSPROFILEDEF(.CURVE.,'CrossProfile-2',.T.,(0.543,3.5,3.5,0.301),(2.47492598733333,3.141592654,3.141592654,2.47714082675747),('L119','L30','0','R30','R119'));
+#75=IFCPOINTBYDISTANCEEXPRESSION(IFCNONNEGATIVELENGTHMEASURE(30.),4.043,0.,0.,#65);
+#76=IFCOPENCROSSPROFILEDEF(.CURVE.,'CrossProfile-3',.T.,(0.663,3.5,3.5,3.154),(2.47492598733333,3.141592654,3.141592654,3.80836500656817),('L119','L30','0','R30','R119'));
```
(0.663,3.5,3.5,3.154)

3.154 for the road side width doesn't seem to match sketch of intent.

javierNadal replied:
Yes, there is a swap between stations 45 and 60

jmirtsch reviewed (commented):
It's a great test case and it's close.  Please check my comments, it's possible I need to improve my own implementation.  I would note that the example isn't a replica of the landxml data, but a simplification.

I think the schema could be improved to allow the section curve to be relocated with start point not at origin, as I'd suggest the intent is to have the middle of the carriage way coincident elevation with the alignment.  You've offset laterally, but offsetting vertically then causes problems with nominating vertical profiles (pivot point).  

![210624 sections gg 02](https://user-images.githubusercontent.com/791757/123184828-d74ae900-d4d7-11eb-9524-34f7d9c60a97.png)

jmirtsch commented on SectionedSurface-Aplitop-1/SectionedSurface-Aplitop-1.IFC:
```diff
@@ -0,0 +1,92 @@
+#67=IFCPRODUCTDEFINITIONSHAPE($,$,(#66));
+#68=IFCFACILITYPART('2VJRykc916eO0F0TUHiUj9',$,$,$,$,#24,#87,$,.COMPLEX.,IFCROADPARTTYPEENUM(.ROADSEGMENT.),.LONGITUDINAL.);
+#69=IFCRELAGGREGATES('0nVIlPiJ1D0Opi4EVcPTpM',#9,$,$,#28,(#68));
+#70=IFCOPENCROSSPROFILEDEF(.CURVE.,'CrossProfile-1',.T.,(0.0310000000000001,3.5,3.5,0.056),(2.49643136367755,3.141592654,3.141592654,2.48087836828573),('L119','L30','0','R30','R119'));
```
![aplitop](https://user-images.githubusercontent.com/791757/123186115-cb145b00-d4da-11eb-8120-48d353acf3d7.png)

Image above might be an incorrect interpretation, but as the last slope ( 2.48087836828573 ) is less than pi, the road side slopes as nominated in ifc would look more like the image below as I understand it.  Can you please confirm?

![210624 sections gg](https://user-images.githubusercontent.com/791757/123186198-0151da80-d4db-11eb-87ad-e00d44d65eef.png)

javierNadal replied:
Yes, slope was wrong

javierNadal committed:
Some fixed errors
